### PR TITLE
Convert path to absolute path in reread,  track all the operations in tree's changes

### PR DESCRIPTION
**Pull request by arthirm**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -405,13 +407,41 @@ FSTree.prototype.match = function(globs) {
...
 FSTree.prototype.changes = function() {
-  return this._changes.filter(change => {
-    return filterMatches(change[1], this.cwd, this.files, this.include, this.exclude);
-  }).map(change => {
-    return [change[0], change[1].replace(`${this.cwd}`, ''), change[2]];
-  });
+  if(this.srcTree && this._changes.length === 0) {
+    if(!this.__hasEntries) {
+      this.entries = walkSync.entries(this.root);
+    }
```
Comment (step2yeung) [inline lib/index.js]: Should we just call this._ensureEntriesPopulated();

File: lib/index.js [hunk 2]
```diff
@@ -405,13 +405,37 @@ FSTree.prototype.match = function(globs) {
 };
 
 FSTree.prototype.changes = function() {
-  return this._changes.filter(change => {
-    return filterMatches(change[1], this.cwd, this.files, this.include, this.exclude);
-  }).map(change => {
-    return [change[0], change[1].replace(`${this.cwd}`, ''), change[2]];
-  });
+  if(this.srcTree && this._changes.length === 0) {
```
Comment (arthirm) [inline lib/index.js]: i think checking for srcTree==true is enough rite?

File: lib/index.js [hunk 3]
```diff
@@ -405,13 +405,37 @@ FSTree.prototype.match = function(globs) {
...
+    let operation;
+    let path = entry.basePath + entry.relativePath;
+
+    if(fs.lstatSync(path).isDirectory()) {
+      operation = "mkdir";
+    } else if(fs.lstatSync(path).isFile()) {
+      operation = "create";
+    } else {
+      throw new Error(`ENOENT: no such file or directory, ${path}`);
+    }
```
Comment (hjdivad) [inline lib/index.js]: we do not want to use `fs` here;   use `isFile` and `isDirectory`, imported at the top of this file.

File: lib/index.js (hunk 2 above)
Comment (hjdivad) [inline lib/index.js]: yes we should only check `this.srcTree`

File: lib/index.js [hunk 4]
```diff
@@ -635,6 +658,10 @@ FSTree.prototype.symlinkSync = function(target, relativePath /*, type */) {
   aware, which is why it is an error to change the root of a non-source tree.
 */
 FSTree.prototype.reread = function(newRoot) {
+  if(newRoot && path.resolve(newRoot) !== newRoot) {
+    newRoot =  path.resolve(newRoot);
+  }
```
Comment (hjdivad) [inline lib/index.js]: i don't think we need to `path.resolve` twice.  it seems clearer to just

```js
if(newRoot) { newRoot = path.resolve(newRoot) + path.sep; }
```

we should then be able to drop the subsequent calls to `normalize` as well as `resolve` will also `normalize`

**Pull request closed without merging**